First pass at adding wording for execution::spawn (#38)
ispeters/async_scope: asyncscope.md

## `execution::spawn`

Add the following as a new subsection immediately after __[exec.nest]__:

::: add
__`std::execution::spawn` [exec.scope.spawn]__

[1]{.pnum} `spawn` associates the given input sender with the given token's async scope and, on success, eagerly starts the input sender.

[2]{.pnum} The name `spawn` denotes a customization point object. For subexpressions `sndr`, `token`, and `env`, let `Sndr` be 

> `decltype((sndr))`, let `Token` be `decltype((token)), and let `Env` be `decltype((env))`. If `sender<Sndr>` or `async_scope_token<Snder>` is false, 

Reviewer: `Snder` should be `Token`.

Reviewer 2: ```suggestion
`decltype((sndr))`, let `Token` be `decltype((token)), and let `Env` be `decltype((env))`. If `sender<Sndr>` or `async_scope_token<Token>` is false, 
```

the expression `spawn(sndr, token, env)` is ill-formed. 

> // TODO: when/how do I specify that env is optional?  Or do we get the env from sender-in?

Reviewer: [3]{.pnum} The expression `spawn(sndr, token)` is expression-equivalent to `spawn(sndr, token, empty_env{})`.

> `[3]{.pnum} The exposition-only class template `impls-for` ([exec.snd.general])  is specialized for `spawn_t` as follows:

Reviewer: `spawn_t` needs to be defined somewhere.

## `execution::spawn`

Add the following as a new subsection immediately after __[exec.nest]__:

::: add
__`std::execution::spawn` [exec.scope.spawn]__

> [1]{.pnum} `spawn` associates the given input sender with the given token's async scope and, on success, eagerly starts the input sender.

Reviewer 2: Since the association could fail, maybe this should say it "attempts" to associate…?
```suggestion
[1]{.pnum} `spawn` attempts to associate the given input sender with the given token's async scope and, on success, eagerly starts the input sender.
```

[2]{.pnum} The name `spawn` denotes a customization point object. For subexpressions `sndr`, `token`, and `env`, let `Sndr` be 

> `decltype((sndr))`, let `Token` be `decltype((token)), and let `Env` be `decltype((env))`. If `sender<Sndr>` or `async_scope_token<Snder>` is false, 

Reviewer 2: I think the formatting is a bit wonky because there's a missing backtick here.
```suggestion
`decltype((sndr))`, let `Token` be `decltype((token))`, and let `Env` be `decltype((env))`. If `sender<Sndr>` or `async_scope_token<Snder>` is false, 
```

the expression `spawn(sndr, token, env)` is ill-formed. 

// TODO: when/how do I specify that env is optional?  Or do we get the env from sender-in?

> `[3]{.pnum} The exposition-only class template `impls-for` ([exec.snd.general])  is specialized for `spawn_t` as follows:

Reviewer 2: I think _`impls-for`_ is a helper for defining a sender and its operation state, so I don't think we need it for `spawn`; instead, we need to define the _`spawn-receiver`_ and _`spawn-env`_ expo-only types and explain what `spawn` does with them.

- otherwise if the expression `get_allocator(get_env(@_new-sender_@))` is well-defined, then `alloc` is the result of `get_allocator(get_env(@_new-sender_@))`
  and `senv` is the expression `@_JOIN-ENV_@(env, @_MAKE-ENV_@(get_allocator, alloc))`
- otherwise `alloc` is `std::allocator<void>{}` and `senv` is the expression `env`

[4]{.pnum} Let _`spawn-state-base`_ be an expositon only class defined below:

```cpp
namespace std::execution {
struct @_spawn-state-base_@ { // exposition-only
    virtual void complete() = 0; // exposition-only
};
}
```

[5]{.pnum} Let _`spawn-receiver`_ be an exposition only class defined below:
```cpp
namespace std::execution {
struct @_spawn-receiver_@ { // exposition-only
    @_spawn-state-base_@* state; // exposition-only
    void set_value() && noexcept { state->complete(); }
    void set_stopped() && noexcept { state->complete(); }
};
}
```

[6]{.pnum} Let_`spawn-state`_ be an exposition only class template defined
below:

```cpp
namespace std::execution {
template<class Alloc, async_scope_token Token, sender Sender>
struct @_spawn-state_@ : @_spawn_state_base_@ {
    using Op = decltype(connect(declval<Sender>(), spawn-receiver{nullptr}));

>     @_spawn-state_@(Alloc alloc, Sender sndr, Token token); // see below

Author: Section 6 - 9 are inspired by split in P2300: https://eel.is/c++draft/exec#split-5

I'm leaving in section 10 for now in case section 6-9 needs to be redone.
